File: abstractions/dotnet/src/ApiClientBuilder.cs
// ------------------------------------------------------------------------------
//  Copyright (c) Microsoft Corporation.  All Rights Reserved.  Licensed under the MIT License.  See License in the project root for license information.
// ------------------------------------------------------------------------------

using System;
using System.Linq;
using Microsoft.Kiota.Abstractions.Serialization;
using Microsoft.Kiota.Abstractions.Store;

namespace Microsoft.Kiota.Abstractions
{
    /// <summary>
    ///     Provides a builder for creating an ApiClient and register the default serializers/deserializers.
    /// </summary>
    public static class ApiClientBuilder
    {
        /// <summary>
        /// Registers the default serializer to the registry.
        /// </summary>
        /// <typeparam name="T">The type of the serialization factory to register</typeparam>
        public static void RegisterDefaultSerializer<T>() where T : ISerializationWriterFactory, new()
        {
            var serializationWriterFactory = new T();
            SerializationWriterFactoryRegistry.DefaultInstance
                                            .ContentTypeAssociatedFactories
                                            .TryAdd(serializationWriterFactory.ValidContentType, serializationWriterFactory);
        }
        /// <summary>
        /// Registers the default deserializer to the registry.
        /// </summary>
        /// <typeparam name="T">The type of the parse node factory to register</typeparam>
        public static void RegisterDefaultDeserializer<T>() where T : IParseNodeFactory, new()
        {
            var deserializerFactory = new T();
            ParseNodeFactoryRegistry.DefaultInstance
                                    .ContentTypeAssociatedFactories
                                    .TryAdd(deserializerFactory.ValidContentType, deserializerFactory);
        }
        /// <summary>
        /// Enables the backing store on default serialization writers and the given serialization writer.
        /// </summary>
        /// <param name="original">The serialization writer to enable the backing store on.</param>
        /// <returns>A new serialization writer with the backing store enabled.</returns>
        public static ISerializationWriterFactory EnableBackingStoreForSerializationWriterFactory(ISerializationWriterFactory original)
        {
            ISerializationWriterFactory result = original ?? throw new ArgumentNullException(nameof(original));
            if(original is SerializationWriterFactoryRegistry registry)
                EnableBackingStoreForSerializationRegistry(registry);
            else
                result = new BackingStoreSerializationWriterProxyFactory(original);
            EnableBackingStoreForSerializationRegistry(SerializationWriterFactoryRegistry.DefaultInstance);
            return result;
        }
        /// <summary>
        /// Enables the backing store on default parse nodes factories and the given parse node factory.
        /// </summary>
        /// <param name="original">The parse node factory to enable the backing store on.</param>
        /// <returns>A new parse node factory with the backing store enabled.</returns>
        public static IParseNodeFactory EnableBackingStoreForParseNodeFactory(IParseNodeFactory original)
        {
            IParseNodeFactory result = original ?? throw new ArgumentNullException(nameof(original));
            if(original is ParseNodeFactoryRegistry registry)
                EnableBackingStoreForParseNodeRegistry(registry);
            else
                result = new BackingStoreParseNodeFactory(original);
            EnableBackingStoreForParseNodeRegistry(ParseNodeFactoryRegistry.DefaultInstance);
            return result;
        }
        private static void EnableBackingStoreForParseNodeRegistry(ParseNodeFactoryRegistry registry)
        {
            foreach(var entry in registry
                                    .ContentTypeAssociatedFactories
                                    .Where(x => !(x.Value is BackingStoreParseNodeFactory ||
                                                    x.Value is ParseNodeFactoryRegistry)))
            {
                registry.ContentTypeAssociatedFactories[entry.Key] = new BackingStoreParseNodeFactory(entry.Value);
            }
        }
        private static void EnableBackingStoreForSerializationRegistry(SerializationWriterFactoryRegistry registry)
        {
            foreach(var entry in registry
                                    .ContentTypeAssociatedFactories
                                    .Where(x => !(x.Value is BackingStoreSerializationWriterProxyFactory ||
                                                    x.Value is SerializationWriterFactoryRegistry)))
            {
                registry.ContentTypeAssociatedFactories[entry.Key] = new BackingStoreSerializationWriterProxyFactory(entry.Value);
            }
        }
    }
}

File: abstractions/typescript/src/apiClientBuilder.ts
import {
  ParseNodeFactory,
  ParseNodeFactoryRegistry,
  SerializationWriterFactory,
  SerializationWriterFactoryRegistry,
} from "./serialization";
import {
  BackingStoreParseNodeFactory,
  BackingStoreSerializationWriterProxyFactory,
} from "./store";

/**
 * Registers the default serializer to the registry.
 * @param type the class of the factory to be registered.
 */
export function registerDefaultSerializer(
  type: new () => SerializationWriterFactory
): void {
  if (!type) throw new Error("Type is required");
  const serializer = new type();
  SerializationWriterFactoryRegistry.defaultInstance.contentTypeAssociatedFactories.set(
    serializer.getValidContentType(),
    serializer
  );
}
/**
 * Registers the default deserializer to the registry.
 * @param type the class of the factory to be registered.
 */
export function registerDefaultDeserializer(
  type: new () => ParseNodeFactory
): void {
  if (!type) throw new Error("Type is required");
  const deserializer = new type();
  ParseNodeFactoryRegistry.defaultInstance.contentTypeAssociatedFactories.set(
    deserializer.getValidContentType(),
    deserializer
  );
}
/**
 * Enables the backing store on default serialization writers and the given serialization writer.
 * @param original The serialization writer to enable the backing store on.
 * @return A new serialization writer with the backing store enabled.
 */
export function enableBackingStoreForSerializationWriterFactory(
  original: SerializationWriterFactory
): SerializationWriterFactory {
  if (!original) throw new Error("Original must be specified");
  let result = original;
  if (original instanceof SerializationWriterFactoryRegistry) {
    enableBackingStoreForSerializationRegistry(
      original as SerializationWriterFactoryRegistry
    );
  } else {
    result = new BackingStoreSerializationWriterProxyFactory(original);
  }
  enableBackingStoreForSerializationRegistry(
    SerializationWriterFactoryRegistry.defaultInstance
  );
  enableBackingStoreForParseNodeRegistry(
    ParseNodeFactoryRegistry.defaultInstance
  );
  return result;
}
/**
 * Enables the backing store on default parse node factories and the given parse node factory.
 * @param original The parse node factory to enable the backing store on.
 * @return A new parse node factory with the backing store enabled.
 */
export function enableBackingStoreForParseNodeFactory(
  original: ParseNodeFactory
): ParseNodeFactory {
  if (!original) throw new Error("Original must be specified");
  let result = original;
  if (original instanceof ParseNodeFactoryRegistry) {
    enableBackingStoreForParseNodeRegistry(
      original as ParseNodeFactoryRegistry
    );
  } else {
    result = new BackingStoreParseNodeFactory(original);
  }
  enableBackingStoreForParseNodeRegistry(
    ParseNodeFactoryRegistry.defaultInstance
  );
  return result;
}
function enableBackingStoreForParseNodeRegistry(
  registry: ParseNodeFactoryRegistry
): void {
  for (const [k, v] of registry.contentTypeAssociatedFactories) {
    if (
      !(
        v instanceof BackingStoreParseNodeFactory ||
        v instanceof ParseNodeFactoryRegistry
      )
    ) {
      registry.contentTypeAssociatedFactories.set(
        k,
        new BackingStoreParseNodeFactory(v)
      );
    }
  }
}
function enableBackingStoreForSerializationRegistry(
  registry: SerializationWriterFactoryRegistry
): void {
  for (const [k, v] of registry.contentTypeAssociatedFactories) {
    if (
      !(
        v instanceof BackingStoreSerializationWriterProxyFactory ||
        v instanceof SerializationWriterFactoryRegistry
      )
    ) {
      registry.contentTypeAssociatedFactories.set(
        k,
        new BackingStoreSerializationWriterProxyFactory(v)
      );
    }
  }
}
